startup.krun

```diff
@@ -0,0 +1,162 @@
+from krun import EntryPoint
+
+# this file is a minimally hacky way to do startup measurements with all the
+# krun goodies, but without changing krun.
+#
+#
+# the way it works is as follows:
+#
+# before the invocation of the VM, another executable is inserted:
+#
+# sudo -u krun nice -20 taskset 0x2 env <some variables> \
+# startup_runner_runner_c jruby startup_runner.rb
+#
+# startup_runner_runner_c is a small C program that prints the current time,
+# then execs its arguments. startup_runner.rb will then also print the current
+# time and exit. The warmup time is therefore the difference of the two (both
+# of these times are stored in the JSON file, the substraction happens in
+# post-processing). That way, all the extra commands are not included in the
+# measurement, but we still get the krun goodies.
+#
+# this works by patching two things:
+# - the runner file for each VM is patched to point to the startup runner
+#   instead (function patch_runner). The startup runners are in
+#   startup_runners/
+# - the bench_cmdline_adjust method is patched to insert the call to
+#   startup_runner_runner_c
+#
+
+# Who to mail
+MAIL_TO = []
+
+# Maximum number of error emails to send per-run
+#MAX_MAILS = 2
+
+DIR = os.getcwd()
+JKRUNTIME_DIR = os.path.join(DIR, "krun", "libkruntime", "")
+JDK8_HOME = os.path.join(DIR, "work/openjdk/build/linux-x86_64-normal-server-release/images/j2sdk-image/")
+STARTUP_RUNNER_DIR = os.path.join(DIR, "startup_runners")
+
+HEAP_LIMIT = 2097152  # K == 2Gb
+
+# Variant name -> EntryPoint
+VARIANTS = {
+    "default-c": EntryPoint("bench.so", subdir="c"),
+    "default-java": EntryPoint("KrunEntry", subdir="java"),
+    "default-lua": EntryPoint("bench.lua", subdir="lua"),
+    "default-python": EntryPoint("bench.py", subdir="python"),
+    "default-php": EntryPoint("bench.php", subdir="php"),
+    "default-ruby": EntryPoint("bench.rb", subdir="ruby"),
+    "default-javascript": EntryPoint("bench.js", subdir="javascript"),
+}
+
+ITERATIONS_ALL_VMS = 1
+
+def patch_runner(vm_def):
+    from krun.env import EnvChangeAppend
+    if vm_def.iterations_runner == "IterationsRunner":
+        #import pdb; pdb.set_trace()
+        vm_def.iterations_runner = "startup_runner"
+        change = EnvChangeAppend("CLASSPATH", STARTUP_RUNNER_DIR)
+        vm_def.add_env_change(change)
+    elif vm_def.iterations_runner.endswith("iterations_runner_c"):
+        vm_def.iterations_runner = os.path.join(STARTUP_RUNNER_DIR, "startup_runner_c")
+    else:
+        assert "iterations_runner." in vm_def.iterations_runner
+        _, suffix = vm_def.iterations_runner.rsplit(".", 1)
+        vm_def.iterations_runner = os.path.join(STARTUP_RUNNER_DIR, "startup_runner." + suffix)
+    return vm_def
+
+from krun.platform import DebianLinuxPlatform, BasePlatform
+
+def bench_cmdline_adjust(self, args, env_dct):
+    prepend_args = BasePlatform.bench_cmdline_adjust(self, [], env_dct)
+    if "check_user_change" in args[-3]:
```

Comment:
The basic ideas is that all the platform checks are run as benchmarks krun. That means their output needs to be valid JSON. This does not work with the startup_runner_runner_c for the check_user_change platform check.